solrmonitor/zkdispatcher.go

```diff
@@ -83,20 +82,13 @@ func NewZkDispatcher(logger zk.Logger) *ZkDispatcher {
 }
 
 func (d *ZkDispatcher) Close() {
-	if !atomic.CompareAndSwapInt32(&d.closed, 0, 1) {
-		panic("already closed")
-	}
-
 	close(d.closedChan)
 }
 
 var errClosed = errors.New("already closed")
 
 // Watch a new ZK event using the given callback to handle the event.
 func (d *ZkDispatcher) WatchEvent(watcher <-chan zk.Event, handler ZkEventCallback) error {
-	if atomic.LoadInt32(&d.closed) != 0 {
```

Comment:
This can allow a new watcher to be added even if the dispatcher has been closed. Maybe first a non-blocking select with *only* the close channel?